mac/Keyman4MacIM/Keyman4MacIM/KMContext.m

```diff
@@ -0,0 +1,201 @@
+/**
+ * Keyman is copyright (C) SIL International. MIT License.
+ * 
+ * KMContext.m
+ * Keyman
+ * 
+ * Created by Shawn Schantz on 2023-05-11.
+ * 
+ * Tracks the state of the context in the client application.
+ * Needed for storing and managing marker actions returned from Keyman Core.
+ * Also useful for applications that do not allow us to read the context.
+ */
+
+#import "KMContext.h"
+#include <Carbon/Carbon.h> /* For kVK_ constants. */
+
+NSUInteger const MAXIMUM_CONTEXT_LENGTH = 80;         // KM_KBP_IT_CHAR
+
+@interface KMContext()
+
+@property (readonly) NSMutableString *context;
+@property BOOL invalid;
+
+@end
+
+@implementation KMContext
+
+-(instancetype)init {
+  self = [super init];
+  if (self) {
+    _context = [[NSMutableString alloc] initWithCapacity:MAXIMUM_CONTEXT_LENGTH];
+    _invalid = YES;
+  }
+  return self;
+}
+
+-(instancetype)initWithString:(NSString*)initialContext {
+  self = [self init];
+  if (self) {
+    [_context setString:[KMContext trimToMaximumContext:initialContext]];
+    _invalid = NO;
+  }
+  return self;
+}
+
++(NSString*)trimToMaximumContext:(NSString*)initialContext {
+  NSString *maximumContext = nil;
+
+  // copy up to MAXIMUM_CONTEXT_LENGTH from the end of the context
+  NSUInteger initialLength = [initialContext length];
+  if (initialLength > MAXIMUM_CONTEXT_LENGTH) {
+    maximumContext = [initialContext substringFromIndex:initialLength-MAXIMUM_CONTEXT_LENGTH];
+  } else {
+    maximumContext = initialContext;
+  }
+
+  return maximumContext;
+}
+
+-(void)resetContext:(NSString*)newContext {
+  [_context setString:[KMContext trimToMaximumContext:newContext]];
+  self.invalid = NO;
+  NSLog(@"   ***CachedContext resetContext, resulting context = %@", _context);
```

Comment:
These `NSLog` functions should be enabled for debug only. Preference is a global debug logging function so we don't scatter `#ifdef DEBUG` everywhere.

Reply:
I can use the function tied to the AppDelegate:
```
  if (self.AppDelegate.debugMode) {
    NSLog(@"containedInNoncompliantAppLists: for app %@: %@", clientAppId, isAppNonCompliant?@"yes":@"no");
  }
```

Reply:
Is it possible to turn that into a helper function so we can keep the log calls to a single line?